src/Database/LSMTree/Internal.hs

```diff
@@ -868,13 +869,16 @@ addRunToLevels ::
   -> HasBlockIO m h
   -> SessionRoot
   -> UniqCounter m
-  -> Run (Handle h)
+  -> Run (PrimState m) (Handle h)
   -> TempRegistry m
-  -> Levels (Handle h)
-  -> m (Levels (Handle h))
+  -> Levels (PrimState m) (Handle h)
+  -> m (Levels (PrimState m) (Handle h))
 addRunToLevels conf@TableConfig{..} resolve hfs hbio root uniqC r0 reg levels = do
     ls' <- go (LevelNo 1) (V.singleton r0) levels
-    assert (runST $ levelsInvariant conf ls') $ return ls'
+#ifdef NO_IGNORE_ASSERTS
+    void $ stToIO $ _levelsInvariant conf ls'
+#endif
+    return ls'
```

Comment:
Interesting, what is the advantage of using CPP here?

Is it because the invariant is not really a predicate (always returning `True`), so it would be weird to use `assert`? In that case, we could also change it to return `()` now and get rid of the `void`.

Reply:
The current `levelsInvariant` doesn't actually read mutable variables, so we can run it purely, but once we add ongoing merges we'll have to read mutable variables to inspect the state of an ongoing merge. As such, we can't use `runST` anymore, and we have to instead embed the invariant check into `IO`

Reply:
And reading the variable won't be possible in ST, like with `MutVar`? Okay, we'll see what it looks like then.

Reply:
The problem is that `runST` is polymorphic over `s`, while in this `addRunToLevels` function we have `s ~ RealWorld` (or more precisely, `s ~ PrimState m`), so we have to embed using `stToIO`